### PR TITLE
PHRAS-3921 prod - expose-cli - became compatible with Phrasea V3 keycloak and fix

### PR DESCRIPTION
**Pull request by aynsix**
### Adds
  - PHRAS-3921: prod - expose-cli - became compatible with Phrasea V3 keycloak and fix

### CODE REVIEW

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 1]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
+                if (isset($refreshtokenBody['refresh_expires_in'])) {
+                    $passSessionNameValue = [
+                        'access_token' => $refreshtokenBody['access_token'],
+                        'expires_at'   => time() + $refreshtokenBody['expires_in'],
+                        'refresh_token'=> $refreshtokenBody['refresh_token'],
+                        'refresh_expires_at' => time() + $refreshtokenBody['refresh_expires_in']
+                    ];
+                } else {
+                    $passSessionNameValue = [
+                        'access_token' => $refreshtokenBody['access_token'],
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1351]: Also set `expires_at`

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 2]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
-                $response = $oauthClient->post($config['expose_base_uri'] . '/oauth/v2/token', [
-                    'json' => [
-                        'client_id'     => $config['expose_client_id'],
-                        'client_secret' => $config['expose_client_secret'],
-                        'grant_type'    => 'client_credentials',
-                        'scope'         => 'publish'
-                    ]
-                ]);
+            if (!isset($tokenInfo['expires_at'])) {
+                $accessToken = $tokenInfo['access_token'];
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1328]: Should never happen

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 3]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
+            } elseif (is_array($tokenInfo) && $tokenInfo['expires_at'] > time()) {
+                $accessToken = $tokenInfo['access_token'];
+            } elseif (is_array($tokenInfo) && $tokenInfo['expires_at'] <= time() && $tokenInfo['refresh_expires_at'] > time()) {
+                $resToken = $this->refreshToken($oauthClient, $config, $tokenInfo['refresh_token']);
 
-                if ($response->getStatusCode() !== 200) {
+                if ($resToken->getStatusCode() !== 200) {
                     return null;
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1335]: Error should be handled here

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 4]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
+                $accessToken = $refreshtokenBody['access_token'];
+            } else {
+                $session->remove($passSessionName);
+            }
 
-                $accessToken = $tokenBody['access_token'];
+        } elseif ($config['connection_kind'] == 'client_credentials') {
+            if ($session->has($credentialSessionName)) {
+                $tokenInfoCredential = $session->get($credentialSessionName);
+                if (!isset($tokenInfoCredential['expires_at'])) {
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1365]: Should always be defined

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 5]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
+    private function getTokenByCredential(Client $oauthClient, array $exposeConfiguration, $credentialSessionName)
+    {
+        $session = $this->getSession();
+
+        $response = $oauthClient->post($exposeConfiguration['oauth_token_uri'] , [
+            'form_params' => [
+                'client_id'     => $exposeConfiguration['expose_client_id'],
+                'client_secret' => $exposeConfiguration['expose_client_secret'],
+                'grant_type'    => 'client_credentials',
+//                'scope'         => 'publish'
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1389]: Does it work if you uncomment?

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 6]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
+            ]
+        ]);
+
+        if ($response->getStatusCode() !== 200) {
+            return null;
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1394]: Error handling

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 7]
```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
...
+        if (isset($tokenBody['expires_in'])) {
+            $credentialSessionNameValue = [
+                'access_token' => $tokenBody['access_token'],
+                'expires_at'   => time() + $tokenBody['expires_in'],
+            ];
+        } else {
+            $credentialSessionNameValue = [
+                'access_token' => $tokenBody['access_token'],
+            ];
+        }
```
Comment (4rthem) [inline lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:1410]:
```suggestion
        $credentialSessionNameValue = [
            'access_token' => $tokenBody['access_token'],
            'expires_at'   => time() + $tokenBody['expires_in'],
        ];
```

**Review by 4rthem — CHANGES_REQUESTED** [review]

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php [hunk 8]
```diff
@@ -80,7 +72,20 @@ public function authenticateAction(PhraseaApplication $app, Request $request)
...
+        if (isset($tokenBody['refresh_expires_in'])) {
+            $passSessionNameValue = [
+                'access_token' => $tokenBody['access_token'],
+                'expires_at'   => time() + $tokenBody['expires_in'],
+                'refresh_token'=> $tokenBody['refresh_token'],
+                'refresh_expires_at' => time() + $tokenBody['refresh_expires_in']
+            ];
+        } else {
+            $passSessionNameValue = [
+                'access_token' => $tokenBody['access_token'],
```
[the remark this answers is not in the document]
Comment (aynsix) [reply lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php:84]: ok

File: lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php (hunk 5 above)
Comment (aynsix) [reply lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php]: yes for v2 and v3

Comment (4rthem) [reply lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php]: OK so uncomment it :) so this client will explicitly request this scope.

**Pull request merged**